File: q51-75/q63.py
# 2020.08.30
# maybe won't do leetcode tomorrow
# Problem Statement:
# https://leetcode.com/problems/unique-paths-ii/

class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m = len(obstacleGrid[0])
        n = len(obstacleGrid)
        
        # check corner cases and do early return
        if m == 1 or n == 1:
            s = 0
            for i in range(0, n):
                for j in range(0, m):
                    if obstacleGrid[i][j] == 1:
                        return 0
            return 1
        
        # initialize the answer
        answer = [[0 for i in range(m)] for j in range(n)]

        # do it line by line
        for i in range(0, n):
            for j in range(0, m):
                # can not reach the obstacle, therefore 0 is filled in
                if obstacleGrid[i][j] == 1:
                    answer[i][j] = 0
                # side values, depend on their neighbour
                elif (i == 0 and j == 0) or \
                     (i == 0 and j>=1 and answer[0][j-1] != 0) or \
                     (j == 0 and i>=1 and answer[i-1][j] != 0): 
                    answer[i][j] = 1
                else:
                    answer[i][j] = answer[i][j-1] + answer[i-1][j]
                
        return answer[n-1][m-1]

File: Amazon/q63.java
// 2022.06.02
// Problem Statement:
// https://leetcode.com/problems/unique-paths-ii/

// idea: dp, from destination to start point
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int [][] dp = new int [obstacleGrid.length][obstacleGrid[0].length];
        // base case vertical
        for (int i=obstacleGrid.length-1; i>=0; i--) {
            if (obstacleGrid[i][obstacleGrid[0].length-1]==1) {
                dp[i][obstacleGrid[0].length-1] = 0;
            } else if (i<obstacleGrid.length-1 && dp[i+1][obstacleGrid[0].length-1]==0) {
                dp[i][obstacleGrid[0].length-1] = 0;
            } else {
                dp[i][obstacleGrid[0].length-1] = 1;
            }
        }
        // base case horizontal
        for (int i=obstacleGrid[0].length-1; i>=0; i--) {
            if (obstacleGrid[obstacleGrid.length-1][i]==1) {
                dp[obstacleGrid.length-1][i] = 0;
            } else if (i<obstacleGrid[0].length-1 && dp[obstacleGrid.length-1][i+1]==0) {
                dp[obstacleGrid.length-1][i] = 0;
            } else {
                dp[obstacleGrid.length-1][i] = 1;
            }
        }
        
        for (int i=obstacleGrid.length-2; i>=0; i--) {
            for (int j=obstacleGrid[0].length-2; j>=0; j--) {
                if (obstacleGrid[i][j]==1) {
                    dp[i][j] = 0;
                } else {
                    dp[i][j] = dp[i+1][j] + dp[i][j+1];
                }
            }
        }
        return dp[0][0];
    }
}